wolfcrypt/src/logging.c

```diff
@@ -481,274 +430,518 @@ WOLFSSL_API int WOLFSSL_IS_DEBUG_ON(void)
+static struct wc_error_queue* wc_current_node;
+
+/* heap info currently used for allocation of entries */
+static void* wc_error_heap;
+
+/* mutex for list operation protection */
+static wolfSSL_Mutex wc_error_mutex;
+#define ERRQ_MUTEX_INIT()      wc_InitMutex(&wc_error_mutex)
+#define ERRQ_MUTEX_FREE()      wc_FreeMutex(&wc_error_mutex)
+#define ERRQ_LOCK()            wc_LockMutex(&wc_error_mutex)
+#define ERRQ_UNLOCK()          wc_UnLockMutex(&wc_error_mutex)
+
+/* Internal function that is called by wolfCrypt_Init() */
+int wc_LoggingInit(void)
+{
+    if (ERRQ_MUTEX_INIT() != 0) {
+        WOLFSSL_MSG("Bad Init Mutex");
+        return BAD_MUTEX_E;
+    }
+    wc_errors_count = 0;
+    wc_errors          = NULL;
+    wc_current_node    = NULL;
+    wc_last_node       = NULL;
+    return 0;
+}
+
+
+/* internal function that is called by wolfCrypt_Cleanup */
+int wc_LoggingCleanup(void)
+{
+    /* clear logging entries */
+    wc_ClearErrorNodes();
+    /* free mutex */
+    if (ERRQ_MUTEX_FREE() != 0) {
+        WOLFSSL_MSG("Bad Mutex free");
+        return BAD_MUTEX_E;
+    }
+    return 0;
+}
+
+static int peekErrorNode(int idx, const char **file, const char **reason,
+        int *line)
+{
+    struct wc_error_queue* err;
+
+    if (idx < 0) {
+        err = wc_last_node;
+    }
+    else {
+        int i;
+
+        err = (struct wc_error_queue*)wc_errors;
+        for (i = 0; i < idx; i++) {
+            if (err == NULL) {
+                WOLFSSL_MSG("Error node not found. Bad index?");
+                return BAD_FUNC_ARG;
+            }
+            err = err->next;
+        }
+    }
+
+    if (err == NULL) {
+        WOLFSSL_MSG("No Errors in queue");
+        return BAD_STATE_E;
+    }
+
+    if (file != NULL) {
+        *file = err->file;
+    }
+    if (reason != NULL) {
+        *reason = err->error;
+    }
+    if (line != NULL) {
+        *line = err->line;
+    }
+
+    return err->value;
+}
+
+/* peek at an error node
+ *
+ * idx : if -1 then the most recent node is looked at,
+ *       otherwise search through queue for node at the given index starting
+ *          from the absolute head wc_errors
+ * file  : pointer to internal file string
+ * reason : pointer to internal error reason
+ * line  : line number that error happened at
+ *
+ * Returns a negative value in error case, on success returns the nodes error
+ * value which is positive (absolute value)
+ */
+int wc_PeekErrorNode(int idx, const char **file, const char **reason,
+        int *line)
+{
+    int ret;
+
+    if (ERRQ_LOCK() != 0) {
+        WOLFSSL_MSG("Lock debug mutex failed");
+        return BAD_MUTEX_E;
+    }
+    ret = peekErrorNode(idx, file, reason, line);
+    ERRQ_UNLOCK();
+    return ret;
+}
+
+static int pullErrorNode(const char **file, const char **reason, int *line)
+{
+    struct wc_error_queue* err;
+    int value;
+
+    err = wc_current_node;
+    if (err == NULL) {
+        WOLFSSL_MSG("No Errors in queue");
+        ERRQ_UNLOCK();
```

Comment:
Caller handles `ERRQ_UNLOCK`. This will get called twice.

Reply:
Thanks for catching this. I rebased and squashed.